server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -120,74 +296,357 @@ public void collect(int doc, long bucket) throws IOException {
+     * rounding all of the keys for {@code owningBucketOrd} that we're going to
+     * collect and picking the rounding based on a real, accurate count and the
+     * min and max.
+     */
+    private static class FromMany extends AutoDateHistogramAggregator {
+        /**
+         * An array of prepared roundings in the same order as
+         * {@link #roundingInfos}. The 0th entry is prepared initially,
+         * and other entries are null until first needed.
+         */
+        private final Rounding.Prepared[] preparedRoundings;
+        /**
+         * Map from value to bucket ordinals.
+         * <p>
+         * It is important that this is the exact subtype of
+         * {@link LongKeyedBucketOrds} so that the JVM can make a monomorphic
+         * call to {@link LongKeyedBucketOrds#add(long, long)} in the tight
+         * inner loop of {@link LeafBucketCollector#collect(int, long)}.
+         */
+        private LongKeyedBucketOrds.FromMany bucketOrds;
+        /**
+         * The index of the rounding that each {@code owningBucketOrd} is
+         * currently using.
+         * <p>
+         * During collection we use overestimates for how much buckets are save
+         * by bumping to the next rounding index. So we end up bumping less
+         * aggressively than a "perfect" algorithm. That is fine because we
+         * correct the error when we merge the buckets together all the way
+         * up in {@link InternalAutoDateHistogram#reduceBucket}. In particular,
+         * on final reduce we bump the rounding until it we appropriately
+         * cover the date range across all of the results returned by all of
+         * the {@link AutoDateHistogramAggregator}s. 
+         */
+        private ByteArray roundingIndices;
+        /**
+         * The minimum key per {@code owningBucketOrd}.
+         */
+        private LongArray mins;
+        /**
+         * The max key per {@code owningBucketOrd}.
+         */
+        private LongArray maxes;
+
+        /**
+         * An underestimate of the number of buckets that are "live" in the
+         * current rounding for each {@code owningBucketOrdinal}. 
+         */
+        private IntArray liveBucketCountUnderestimate;
+        /**
+         * An over estimate of the number of wasted buckets. When this gets
+         * too high we {@link #rebucket} which sets it to 0.
+         */
+        private long wastedBucketsOverestimate = 0;
+        /**
+         * The next {@link #wastedBucketsOverestimate} that will trigger a
+         * {@link #rebucket() rebucketing}.
+         */
+        private long nextRebucketAt = 1000; // TODO this could almost certainly start higher when asMultiBucketAggregator is gone
+        /**
+         * The number of times the aggregator had to {@link #rebucket()} the
+         * results. We keep this just to report to the profiler.
+         */
+        private int rebucketCount = 0;
```

Comment:
Exposing rebucketing info (count, rebucketAt, wasted, etc) via `collectDebugInfo` would be super interesting from a profiling perspective.  Especially if a user comes to us with performance issues, we'd have a better idea if they are hitting some kind of antagonistic scenario or something

Not super important, just a thought :)

Reply:
I already exposed it all! I made this in the first place because I wanted to *see* how the algorithm did with certain bits of data with the profiler.

Reply:
Wow I totally overlooked those.  Ignore me! :)